### ISSUE TITLE
Configurable Vault secrets renewal attempt interval

### ISSUE THREAD

**Issue by marcelocarlos** [hashicorp/consul-template#1347]
Hi,

We are currently looking into ways of improving our integration between Vault and Applications in kubernetes. A very common pattern is to use consul-template as a sidecar container and let it handle secrets retrieval and renewal, and it works very well!

However, there are two items we would like to further customise. The first is monitoring (which is a very similar use case of https://github.com/hashicorp/consul-template/issues/1338). The second, which this ticket is about, is having the ability of tune the secrets renewal interval defined in https://github.com/hashicorp/consul-template/blob/5524082455f72a9edd4c92f848807e4a3a5e8f2d/dependency/vault_common.go#L145-L164. 

Ideally, for renewable secrets we'd like to be able to be able to define when (within the lease validity period) the renew attempts would be performed and how many attempts should be done. For example, in a lease valid for 24h, we would like to have 5 attempts to renew the lease, potentially more towards the end of the lease, similar to what's being done to non-renewable secrets. For non-renewable secrets, we would like to customise the number of attempts and the window used.

Assuming this is possible, we could report the attempts in prometheus format so this can be monitored, alerted and acted upon (assuming https://github.com/hashicorp/consul-template/issues/1338 is implemented).

In summary, we could maybe first allow customising the number of attempts (make the current `3` hard-coded attempts a customisable setting for renewable secrets and adding it for non-renewable secrets too). Then also allow specifying when the attempts would take place (allow customising the `.85` hard-coded value in the non-renewable secrets and add it for renewable secrets as well).

Thanks you

**Comment by eikenb** [comment]
Hey @marcelocarlos, thanks for the interest.

We just added support for using `rotation_period` settings on non-renewable secrets. You can see the changes in #1358, which has been merged (only in master for now). Thought I'd mention it in case you would find that useful.

**Comment by phemmer** [comment]
I'm also interested in such a feature, but would like to set a different interval for k/v secrets fetched by version, possibly with the ability to set an "infinite duration" interval. The only possibility for a k/v versioned secret to change is through deletion.  
In my use case, I guarantee a secret will never be deleted/destroyed (consul-template has terrible handling of deleted secrets), so there's no need to ever re-check it.

**Comment by Oloremo** [comment]
Interested in this as well. We need to set renewal time to create some guarantees with our monitoring.
